### DISCUSSION TITLE
[Feature Request] Support ring 0 emulation (kernel emulation)?

### DISCUSSION THREAD

**Discussion by Fubuki0x10DE** [86Box/86Box#4223] [Ideas]
Emulate CPU (i586 / K5 and higher) using kernel driver ("86box_x86.sys" and "86box_x64.sys"). To retain compatibility with Win9x guest, hardware virtualization (VMX and SVM) are not used. This is similar to "binary translation" in Virtual PC and VMware Workstation.
Pros: Good emulate performance under Windows x86 / x64 host, solve sound stutters in Win9x / Win2000 guest.
Cons: This driver needs WHQL signature.

**Comment by OBattler** [comment]
And it would break OPL, as seen in Virtual PC, where OPL sounds as atrocious as it does on real too fast hardware.

Also, Windows 98 and 2000 are already pretty well emulatable right now, you can solve sound stutters by reducing the emulated CPU speed.

**Reply by OBattler** [reply]
Not to mention - it would require a second emulation of each timed device to make it threaded and instant-access like on Virtual PC.

**Reply by lemondrops** [reply]
This doesn't address his points at all...

**Reply by Fubuki0x10DE** [reply]
On our i5-3210M laptop, 86Box can only emulate Pentium 100MHz at 100%. Hope kernel emulation can be an 86Box feature.
For kernel emulation mode, only PCI sound device is allowed. SB PCI and CMedia PCI are don't have physical OPL.

**Reply by TC1995** [reply]
> On our i5-3210M laptop, 86Box can only emulate Pentium 100MHz at 100%. Hope kernel emulation can be an 86Box feature. For kernel emulation mode, only PCI sound device is allowed. SB PCI and CMedia PCI are don't have physical OPL.

If you can write such a driver, then fine, but otherwise, no.

**Reply by ppgrainbow** [reply]
I personally believe that it would be nearly impossible to write a Ring 0 kernel emulation as it would require "secondary emulation of each timed device to make it threaded" and instant access as it would break OPL functionality. That would be too much work for the foreseeable future.

On my main PC, I can only emulate up to either a 66 MHz 80486DX2 or a 66 MHz Pentium at 100%.
